test/e2e_env/universal/gateway/gateway.go

```diff
@@ -0,0 +1,267 @@
+type: ExternalService
+mesh: %s
+name: external-service
+tags:
+  kuma.io/service: external-echo
+networking:
+  address: "%s"
+`, mesh, net.JoinHostPort(env.Cluster.GetApp("external-echo-gateway").GetIP(), "8080"))),
+			).To(Succeed())
+		})
+
+		It("should proxy simple HTTP requests", func() {
+			proxySimpleRequests(env.Cluster, "external-echo",
+				GatewayAddressPort("gateway-proxy", gatewayPort), "example.kuma.io",
+				client.WithPathPrefix("/external"))
+		})
+	})
+	Context("applying ProxyTemplate", func() {
+		It("shouldn't error with gateway-proxy import", func() {
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: ProxyTemplate
+mesh: %s
+name: edge-gateway
+selectors:
+  - match:
+      kuma.io/service: edge-gateway
+conf:
+  imports:
+    - gateway-proxy
+  modifications:
+    - cluster:
+        operation: add
+        value: |
+          name: test-cluster
+          connectTimeout: 5s
+          type: STATIC
+`, mesh)),
+			).To(Succeed())
+		})
+	})
+
+	Context("when a rate limit is configured", func() {
+		BeforeAll(func() {
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: RateLimit
+mesh: %s
+name: echo-rate-limit
+sources:
+- match:
+    kuma.io/service: edge-gateway
+destinations:
+- match:
+    kuma.io/service: echo-service
+conf:
+  http:
+    requests: 5
+    interval: 10s
+`, mesh)),
+			).To(Succeed())
+		})
+
+		It("should be rate limited", func() {
+			proxyRequestsWithRateLimit(env.Cluster,
+				GatewayAddressPort("gateway-proxy", gatewayPort))
+		})
+	})
+
+	Context("when targeting a HTTPS gateway", func() {
+		BeforeAll(func() {
+			// Delete the default gateway that the test fixtures create.
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlDelete("meshgateway", "edge-gateway", mesh),
+			).To(Succeed())
+
+			// And replace it with a HTTPS gateway.
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: MeshGateway
+mesh: %s
+name: edge-https-gateway
+selectors:
+- match:
+    kuma.io/service: edge-gateway
+conf:
+  listeners:
+  - port: 8080
+    protocol: HTTPS
+    hostname: example.kuma.io
+    tls:
+      mode: TERMINATE
+      certificates:
+      - secret: example-kuma-io-certificate
+    tags:
+      hostname: example.kuma.io
+  - port: 8081
+    protocol: HTTPS
+    tls:
+      mode: TERMINATE
+      certificates:
+      - secret: example-kuma-io-certificate
+`, mesh)),
+			).To(Succeed())
+
+			cert, key, err := CreateCertsFor("example.kuma.io")
+			Expect(err).To(Succeed())
+
+			payload := base64.StdEncoding.EncodeToString([]byte(strings.Join([]string{key, cert}, "\n")))
+
+			// Create the TLS secret containing the self-signed certificate and corresponding private key.
+			Expect(
+				env.Cluster.GetKumactlOptions().KumactlApplyFromString(fmt.Sprintf(`
+type: Secret
+mesh: %s
+name: example-kuma-io-certificate
+data: %s
+	`, mesh, payload)),
+			).To(Succeed())
```

Comment:
nits:
1) I think we should create secrets first
2) Can we just override existing gateway? less code
3) `env.Cluster.Install(YamlUniversal(` is less verbose than `env.Cluster.GetKumactlOptions().KumactlApplyFromString(` and more visible around e2e codebase

Reply:
1) ok
2) Sure though a bit conflicted on making everything here order-dependent. My gut feeling says HTTP and HTTPS tests shouldn't depend on each other.
3) Definitely, I didn't actually realize there was an `Install` method, I thought that was a ClusterSetup method only.

Reply:
If I'm ordering them I'll ust do the trafficpermission delete at the end so I can use one mesh.

Reply:
> Sure though a bit conflicted on making everything here order-dependent. My gut feeling says HTTP and HTTPS tests shouldn't depend on each other.

that's a good gut feeling! but don't they depend on the order of each other right now? Would this work if we run `when targeting a HTTPS gateway` first and then `when mTLS is disabled`?

Reply:
In the first commit I don't think the order does matter. I use different meshes for mTLS enabled/disabled.

Reply:
The changes I just pushed fully embraces `Ordered`.